// pizza-delivery-frontend/src/hooks/useSocket.js
import { useEffect, useRef } from 'react'
import { io } from 'socket.io-client'
import { useDispatch, useSelector } from 'react-redux'
import toast from 'react-hot-toast'
import { updateOrderStatus } from '../redux/orderSlice'
import { addNotification } from '../redux/notificationSlice'
import { selectToken, selectIsAuthenticated } from '../redux/authSlice'

const SOCKET_URL = import.meta.env.VITE_API_URL?.replace('/api', '') || 'https://pizza-delivery-zjn3.onrender.com'

export const useSocket = () => {
  const dispatch        = useDispatch()
  const isAuthenticated = useSelector(selectIsAuthenticated)
  const token           = useSelector(selectToken)
  const socketRef       = useRef(null)

  useEffect(() => {
    if (!isAuthenticated || !token) return

    // Connect with auth token
    socketRef.current = io(SOCKET_URL, {
      auth: { token },
      transports: ['websocket'],
    })

    const socket = socketRef.current

    socket.on('connect', () => {
      console.log('Socket connected:', socket.id)
    })

    // Listen for order status updates from admin
    socket.on('orderStatusUpdated', ({ orderId, status }) => {
      dispatch(updateOrderStatus({ orderId, status }))
      toast.success(`Order update: ${status}`, { icon: '🍕' })
      dispatch(addNotification({
        type: 'order',
        title: `Order ${orderId} updated`,
        message: `Status changed to ${status}`,
        data: { orderId, status },
      }))
    })

    // Order placed (real-time)
    socket.on('orderPlaced', (payload) => {
      const { orderId, short } = payload || {}
      toast.success('New order placed', { icon: '🧾' })
      dispatch(addNotification({ type: 'order', title: `New order ${orderId}`, message: short || 'A new order was placed', data: payload }))
    })

    // Order delivered
    socket.on('orderDelivered', (payload) => {
      const { orderId } = payload || {}
      toast.success('Order delivered 🎉', { icon: '🚚' })
      dispatch(addNotification({ type: 'order', title: `Order ${orderId} delivered`, message: 'Order has been delivered', data: payload }))
    })

    // Stock low alert
    socket.on('stockLow', (payload) => {
      const { ingredient, remaining } = payload || {}
      toast.error(`Low stock: ${ingredient}`, { icon: '⚠️' })
      dispatch(addNotification({ type: 'warning', title: `Low stock: ${ingredient}`, message: `Remaining: ${remaining}`, data: payload }))
    })

    // Admin alerts
    socket.on('adminAlert', (payload) => {
      const { title, message } = payload || {}
      toast.info(`${title ? title + ': ' : ''}${message || 'Admin message'}`)
      dispatch(addNotification({ type: 'alert', title: title || 'Admin Alert', message: message || '', data: payload }))
    })

    socket.on('disconnect', () => {
      console.log('Socket disconnected')
    })

    socket.on('connect_error', (err) => {
      console.error('Socket error:', err.message)
    })

    return () => {
      socket.disconnect()
    }
  }, [isAuthenticated, token, dispatch])

  return socketRef.current
}
