// pizza-delivery-frontend/src/pages/user/Checkout.jsx
import React, { useState } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { formatPrice } from '../../utils/helpers'
import { placeOrder } from '../../redux/orderSlice'
import { clearItems } from '../../redux/cartSlice'
import { useNavigate } from 'react-router-dom'
import toast from 'react-hot-toast'

export default function Checkout() {
  const items = useSelector((state) => state.cart.items)
  const dispatch = useDispatch()
  const navigate = useNavigate()
  const [address, setAddress] = useState('')
  const [phone, setPhone] = useState('')
  const [paymentMethod, setPaymentMethod] = useState('cash')
  const [loading, setLoading] = useState(false)
  const [errorMessage, setErrorMessage] = useState('')

  const total = items.reduce((s, i) => s + (i.lineTotal || 0), 0)

  const handlePlaceOrder = async () => {
    setErrorMessage('')
    if (!address.trim() || !phone.trim()) {
      setErrorMessage('Please provide both delivery address and phone number.')
      return
    }

    const orderPayload = {
      pizzas: items.map((i) => ({
        pizza: i.pizzaId,
        name: i.name,
        size: i.size,
        sizePrice: i.sizePrice,
        toppings: (i.toppings || []).map((t) => ({
          ingredientId: t.ingredientId,
          name: t.name,
          extraPrice: t.extraPrice,
        })),
        quantity: i.quantity,
      })),
      address,
      phone,
      paymentMethod,
      totalPrice: total,
    }

    setLoading(true)
    try {
      await dispatch(placeOrder(orderPayload)).unwrap()
      dispatch(clearItems())
      toast.success('Order placed successfully!')
      navigate('/my-orders')
    } catch (err) {
      setErrorMessage(err || 'Order failed. Please try again.')
      toast.error(err || 'Order failed. Please try again.')
    } finally {
      setLoading(false)
    }
  }

  if (!items.length) {
    return (
      <div className="p-6 min-h-[60vh]">
        <div className="mx-auto max-w-2xl rounded-[2rem] border border-dashed border-slate-300 bg-slate-50 p-10 text-center shadow-sm">
          <p className="text-sm uppercase tracking-[0.2em] text-slate-500">Checkout unavailable</p>
          <h2 className="mt-4 text-3xl font-semibold text-slate-900">Your cart is empty</h2>
          <p className="mt-3 text-slate-600">Add items to your cart before placing an order.</p>
          <button
            type="button"
            onClick={() => navigate('/dashboard')}
            className="mt-8 rounded-full bg-primary-600 px-6 py-3 text-sm font-semibold text-white transition hover:bg-primary-700"
          >
            Browse pizzas
          </button>
        </div>
      </div>
    )
  }

  return (
    <div className="p-6 max-w-7xl mx-auto">
      <div className="mb-6 flex flex-col gap-4 md:flex-row md:items-end md:justify-between">
        <div>
          <p className="text-sm uppercase tracking-[0.2em] text-slate-500">Secure checkout</p>
          <h2 className="mt-2 text-3xl font-semibold text-slate-900">Review your order</h2>
        </div>
        <div className="rounded-full bg-slate-100 px-4 py-2 text-sm font-medium text-slate-700">
          {items.length} {items.length === 1 ? 'item' : 'items'} in cart
        </div>
      </div>

      <div className="grid gap-8 lg:grid-cols-[1.4fr_0.75fr]">
        <div className="space-y-6 rounded-[2rem] border border-slate-200 bg-white p-6 shadow-sm">
          <div>
            <h3 className="text-xl font-semibold text-slate-900">Delivery details</h3>
            <p className="mt-2 text-slate-600">Fill in your address and phone number to complete the order.</p>
          </div>

          <div className="grid gap-5">
            <label className="block">
              <span className="text-sm font-medium text-slate-700">Delivery address</span>
              <textarea
                value={address}
                onChange={(e) => setAddress(e.target.value)}
                className="mt-3 w-full rounded-3xl border border-slate-200 bg-slate-50 p-4 text-sm text-slate-900 outline-none transition focus:border-primary-500 focus:bg-white"
                rows={4}
              />
            </label>

            <label className="block">
              <span className="text-sm font-medium text-slate-700">Phone number</span>
              <input
                value={phone}
                onChange={(e) => setPhone(e.target.value)}
                placeholder="123-456-7890"
                className="mt-3 w-full rounded-3xl border border-slate-200 bg-slate-50 p-4 text-sm text-slate-900 outline-none transition focus:border-primary-500 focus:bg-white"
              />
            </label>

            <div>
              <p className="text-sm font-medium text-slate-700">Payment method</p>
              <div className="mt-3 grid gap-3 sm:grid-cols-3">
                {[
                  { value: 'cash', label: 'Cash on Delivery' },
                  { value: 'card', label: 'Card' },
                  { value: 'upi', label: 'UPI' },
                ].map((option) => (
                  <button
                    key={option.value}
                    type="button"
                    onClick={() => setPaymentMethod(option.value)}
                    className={`rounded-3xl border px-4 py-3 text-left text-sm transition ${
                      paymentMethod === option.value
                        ? 'border-primary-600 bg-primary-600 text-white'
                        : 'border-slate-200 bg-white text-slate-700 hover:border-slate-300'
                    }`}
                  >
                    {option.label}
                  </button>
                ))}
              </div>
            </div>

            {errorMessage && (
              <div className="rounded-3xl border border-rose-200 bg-rose-50 p-4 text-sm text-rose-700">
                {errorMessage}
              </div>
            )}

            <button
              type="button"
              onClick={handlePlaceOrder}
              disabled={loading}
              className="w-full rounded-full bg-primary-600 px-6 py-3 text-base font-semibold text-white transition hover:bg-primary-700 disabled:cursor-not-allowed disabled:opacity-70"
            >
              {loading ? 'Placing order...' : 'Place order'}
            </button>
          </div>
        </div>

        <aside className="rounded-[2rem] border border-slate-200 bg-slate-50 p-6 shadow-sm">
          <div className="space-y-5">
            <div>
              <p className="text-sm uppercase tracking-[0.18em] text-slate-500">Order summary</p>
              <h3 className="mt-2 text-2xl font-semibold text-slate-900">Your cart</h3>
            </div>

            <div className="space-y-4 rounded-3xl bg-white p-5 shadow-sm">
              {items.map((item) => (
                <div key={item._key} className="flex items-start justify-between gap-4">
                  <div>
                    <p className="font-semibold text-slate-900">{item.name}</p>
                    <p className="text-sm text-slate-500">
                      {item.size} · Qty {item.quantity}
                    </p>
                  </div>
                  <p className="font-semibold text-slate-900">{formatPrice(item.lineTotal)}</p>
                </div>
              ))}
            </div>

            <div className="rounded-3xl bg-white p-5 shadow-sm">
              <div className="flex items-center justify-between text-sm text-slate-500">
                <span>Subtotal</span>
                <span>{formatPrice(total)}</span>
              </div>
              <div className="mt-4 flex items-center justify-between text-lg font-semibold text-slate-900">
                <span>Total</span>
                <span>{formatPrice(total)}</span>
              </div>
            </div>
          </div>
        </aside>
      </div>
    </div>
  )
}
