// pizza-delivery-frontend/src/pages/pizza/PizzaDetails.jsx
import React, { useEffect, useMemo, useState } from 'react'
import { useParams, useNavigate } from 'react-router-dom'
import { pizzaApi } from '../../api/pizzaApi'
import { formatPrice } from '../../utils/helpers'
import { useDispatch } from 'react-redux'
import { addItem } from '../../redux/cartSlice'
import Loader from '../../components/common/Loader'

export default function PizzaDetails() {
  const { id } = useParams()
  const navigate = useNavigate()
  const dispatch = useDispatch()
  const [pizza, setPizza] = useState(null)
  const [loading, setLoading] = useState(true)
  const [error, setError] = useState('')
  const [size, setSize] = useState('')
  const [toppings, setToppings] = useState([])
  const [quantity, setQuantity] = useState(1)

  useEffect(() => {
    setLoading(true)
    setError('')
    pizzaApi.getPizzaById(id)
      .then((response) => {
        const pizzaData = response.data
        setPizza(pizzaData)
        if (pizzaData.sizes?.length) {
          setSize(pizzaData.sizes[0].name)
        } else {
          setSize(pizzaData.size || 'Regular')
        }
      })
      .catch((err) => {
        setError(err.response?.data?.message || 'Unable to load pizza details. Please try again.')
      })
      .finally(() => setLoading(false))
  }, [id])

  const imageSrc = pizza?.image || pizza?.imageUrl || pizza?.photo
  const selectedSizeObj = useMemo(() => {
    if (pizza?.sizes?.length) {
      return pizza.sizes.find((s) => s.name === size) || pizza.sizes[0]
    }
    return { name: size || 'Regular', price: pizza?.price || 0 }
  }, [pizza, size])

  const toppingsList = pizza?.toppings || []
  const toppingsTotal = toppings.reduce((sum, t) => sum + (t.extraPrice || 0), 0)
  const unitPrice = (selectedSizeObj?.price || 0) + toppingsTotal
  const totalPrice = unitPrice * Math.max(1, quantity)

  const toggleTopping = (topping) => {
    const exists = toppings.some((item) => item.ingredientId === topping.ingredientId)
    if (exists) {
      setToppings(toppings.filter((item) => item.ingredientId !== topping.ingredientId))
    } else {
      setToppings([...toppings, topping])
    }
  }

  const handleAdd = () => {
    dispatch(
      addItem({
        pizzaId: pizza._id,
        name: pizza.name,
        size: selectedSizeObj.name,
        sizePrice: selectedSizeObj.price,
        toppings,
        quantity: Math.max(1, quantity),
      })
    )
    navigate('/cart')
  }

  if (loading) return <Loader fullScreen />

  if (error) {
    return (
      <div className="p-6 max-w-3xl mx-auto">
        <div className="rounded-3xl border border-red-200 bg-red-50 p-8 text-center">
          <p className="text-lg font-semibold text-red-700">{error}</p>
          <button
            type="button"
            onClick={() => navigate(-1)}
            className="mt-6 rounded-full bg-primary-600 px-5 py-2 text-sm font-semibold text-white hover:bg-primary-700"
          >
            Go back
          </button>
        </div>
      </div>
    )
  }

  return (
    <div className="p-6 max-w-6xl mx-auto">
      <button className="mb-6 text-sm text-slate-500 hover:text-slate-900" onClick={() => navigate(-1)}>
        ← Back to menu
      </button>

      <div className="grid gap-10 lg:grid-cols-[1.15fr_0.85fr] items-start">
        <div className="space-y-6 rounded-[2rem] border border-slate-200 bg-white p-6 shadow-sm">
          <div className="overflow-hidden rounded-[1.75rem] bg-slate-100">
            {imageSrc ? (
              <img
                src={imageSrc}
                alt={pizza.name}
                className="h-full w-full object-cover"
              />
            ) : (
              <div className="flex h-72 items-center justify-center text-8xl text-orange-500">
                🍕
              </div>
            )}
          </div>

          <div className="space-y-4">
            <div className="flex flex-wrap items-center gap-3">
              <span className="rounded-full bg-primary-50 px-3 py-1 text-xs font-semibold uppercase tracking-[0.18em] text-primary-700">
                {pizza.category || 'Classic'}
              </span>
              <span className="rounded-full bg-slate-100 px-3 py-1 text-xs font-medium text-slate-600">
                {pizza.sizes?.length ? pizza.sizes.length + ' sizes available' : 'Single size'}
              </span>
            </div>

            <div>
              <h1 className="text-3xl font-semibold text-slate-900">{pizza.name}</h1>
              <p className="mt-3 text-gray-600 leading-7">{pizza.description}</p>
            </div>
          </div>

          <div className="grid gap-6 md:grid-cols-2">
            <div className="rounded-3xl bg-slate-50 p-5">
              <p className="text-sm font-semibold uppercase tracking-[0.18em] text-slate-500">Choose size</p>
              <div className="mt-4 flex flex-wrap gap-3">
                {pizza.sizes?.length ? (
                  pizza.sizes.map((option) => (
                    <button
                      key={option.name}
                      type="button"
                      onClick={() => setSize(option.name)}
                      className={`rounded-2xl border px-4 py-3 text-left text-sm transition ${
                        size === option.name
                          ? 'border-primary-600 bg-primary-600 text-white'
                          : 'border-slate-200 bg-white text-slate-700 hover:border-slate-300'
                      }`}
                    >
                      <div className="font-semibold">{option.name}</div>
                      <div className="mt-1 text-sm text-slate-500">{formatPrice(option.price)}</div>
                    </button>
                  ))
                ) : (
                  <div className="rounded-2xl border border-slate-200 bg-white p-4 text-sm text-slate-600">
                    {formatPrice(pizza.price || 0)}
                  </div>
                )}
              </div>
            </div>

            <div className="rounded-3xl bg-slate-50 p-5">
              <p className="text-sm font-semibold uppercase tracking-[0.18em] text-slate-500">Choose toppings</p>
              <div className="mt-4 grid gap-3 sm:grid-cols-2">
                {toppingsList.length ? (
                  toppingsList.map((option) => {
                    const active = toppings.some((item) => item.ingredientId === option.ingredientId)
                    return (
                      <button
                        key={option.ingredientId}
                        type="button"
                        onClick={() => toggleTopping(option)}
                        className={`rounded-2xl border px-4 py-3 text-left text-sm transition ${
                          active
                            ? 'border-primary-600 bg-primary-600 text-white'
                            : 'border-slate-200 bg-white text-slate-700 hover:border-slate-300'
                        }`}
                      >
                        <div className="font-semibold">{option.name}</div>
                        <div className="mt-1 text-slate-500">+ {formatPrice(option.extraPrice || 0)}</div>
                      </button>
                    )
                  })
                ) : (
                  <p className="text-sm text-slate-500">No extra toppings available for this pizza.</p>
                )}
              </div>
            </div>
          </div>
        </div>

        <aside className="space-y-6 rounded-[2rem] border border-slate-200 bg-white p-6 shadow-sm">
          <div className="space-y-3">
            <div className="flex items-center justify-between gap-3">
              <div>
                <p className="text-sm font-semibold uppercase tracking-[0.18em] text-slate-500">Quantity</p>
              </div>
              <div className="flex items-center gap-2 rounded-full border border-slate-200 bg-slate-100 px-3 py-2">
                <button
                  type="button"
                  onClick={() => setQuantity((cur) => Math.max(1, cur - 1))}
                  className="rounded-full bg-white px-3 py-2 text-lg font-semibold text-slate-700 hover:bg-slate-200"
                >
                  −
                </button>
                <span className="min-w-[2rem] text-center text-lg font-semibold">{quantity}</span>
                <button
                  type="button"
                  onClick={() => setQuantity((cur) => cur + 1)}
                  className="rounded-full bg-white px-3 py-2 text-lg font-semibold text-slate-700 hover:bg-slate-200"
                >
                  +
                </button>
              </div>
            </div>

            <div className="rounded-3xl bg-slate-100 p-4">
              <div className="flex items-center justify-between text-sm text-slate-500">
                <span>Size</span>
                <span>{selectedSizeObj?.name || 'Regular'}</span>
              </div>
              <div className="mt-3 flex items-center justify-between text-lg font-semibold text-slate-900">
                <span>Price</span>
                <span>{formatPrice(unitPrice)}</span>
              </div>
            </div>

            <div className="rounded-3xl bg-primary-600 px-5 py-4 text-white shadow-lg">
              <div className="text-sm uppercase tracking-[0.18em] text-primary-100">Total</div>
              <div className="mt-2 text-3xl font-semibold">{formatPrice(totalPrice)}</div>
            </div>

            <button
              type="button"
              onClick={handleAdd}
              className="w-full rounded-full bg-primary-600 px-6 py-3 text-base font-semibold text-white transition hover:bg-primary-700"
            >
              Add to cart
            </button>
          </div>
        </aside>
      </div>
    </div>
  )
}
